net: ip: Fix checksum calculation/validation for fragmented packets

rlubos opened the pull request:
In case HW checksum offloading was enabled on an interface, the checksum calculation/verification was broken for IP fragmented packets. This PR fixes this:
  * Always calculate/verify checksum for IP fragmented packets, regardless of the HW offloading capability
  * `chksum_done` flag is introduced, to indicate checksum status on the net_pkt, which can be used by drivers to determine whether checksum calculation is needed (unless they detect IP fragments automatically, like `eth_stm32_hal.c`, which has been used to verify the functionality on HW),
  * A set of new test cases was introduced, to verify checksums are calculated properly on for fragmented packets
  * Further cleanup in existing checksum offloading tests cases, to reduce code duplication.

Fixes #60822

rlubos committed:
net: pkt: Add flag indicating checksum status

Add new net_pkt checksum, which indicate checksum status on the packet
(i. e. whether it has already been calculated or not).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: Set a flag when checksum has been computed

Set checksum flag on the net_pkt, when checksum is calculated in
software.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ip: Add option to force checksum calculation

Modify internal L4 protocols APIs, to allow to enforce checksum
calculation, regardless of the checksum HW offloading capability.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ipv4: Calculate checksum before fragmentation

In case the stack has to fragment the IPv4 packet, calculate the
checksum before fragmentation (if haven't done so).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ipv6: Calculate checksum before fragmentation

In case the stack has to fragment the IPv6 packet, calculate the
checksum before fragmentation (if haven't done so).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ip: Set net_context on the final fragment

IPv4/6 fragmentation did not set the net_context pointer on the fragment
packet and in result the send callback registered on net_context was not
called. Therefore, copy the net_context pointer from the original packet
to the final fragment to ensure that the registered callback is called.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ipv6: Remove invalid/unneeded code from fragmentation logic

Skipping both next_hdr_off and last_hdr_off and filling last_hdr
variable doesn't make much sense, as this effectively moves the packet
cursor inside/behind the last (L4) header with no particular meaning.
Plus the last_hdr variable isn't really used anywhere, which kind of
proves the point. Therefore, remove the unused variable and needless
net_pkt operations.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: checksum_offload: Add tests involving IP fragmentation

Add test cases verifying proper checksum calculation involving
IPv4/IPv6 fragmentation.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: checksum_offload: Move net context init to helper function

Move code responsible for net context initialization, repeated over
multiple test cases, to a helper function to reduce code duplication.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: checksum_offload: Move test code into helper function

Further reduce code duplication by moving test code for similar test
cases into a helper function, parametrized based on the test case.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ipv6: Set IP header length on the fragmented packet

IPv6 fragmentation code did not set the IP header field on the
fragment net_pkt.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: pkt: Add function to check if packet was reassembled at IP level

Move the existing code verifying that the net_pkt was reassembled at IP
level to a helper function, as it will be needed in other places as
well. Additionally, add packet family check before accessing union
fields.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: Verify L4 checksum unconditionally for reassembled packets

In case of reassembled IP packets, we cannot rely on checksum
offloading as the drivers/HW has no means to verify L4 checksum before
the fragment is reassembled. Therefore, for such packets, verify L4
checksum in the software unconditionally.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

pdgendt reviewed (dismissed):
Not an expert on this, subsystem changes seem logical, but testing part I haven't reviewed thoroughly.

pdgendt commented:
Would be nice to get feedback from the issue author @krish2718 if this PR fixes it.

krish2718 commented on include/zephyr/net/net_pkt.h:
```diff
@@ -194,6 +194,9 @@ struct net_pkt {
 	uint8_t l2_processed : 1; /* Set to 1 if this packet has already been
 				   * processed by the L2
 				   */
+	uint8_t chksum_done : 1; /* L4 Checksum has already been computed for
```
This is used by L3 (IP) as well, no? also if you want abbreviate `csum` might be better?

rlubos replied:
I think we can drop L4 here, we either calculate either both (L3/L4) checksums or none anyway.
As for the abbreviation, I'd prefer to stick to `chksum`, as it's already widely used in Zephyr, in checksum related routines.

krish2718 commented on include/zephyr/net/net_pkt.h:
```diff
@@ -836,6 +836,20 @@ static inline void net_pkt_set_ipv6_fragment_id(struct net_pkt *pkt,
+{
+	if ((IS_ENABLED(CONFIG_NET_IPV4_FRAGMENT) &&
+	     net_pkt_family(pkt) == AF_INET &&
+	     net_pkt_ipv4_fragment_more(pkt)) ||
```
If its re-assembled won't the more_frags be 0?

rlubos replied:
I was puzzled about this part of code (note, that I've just moved it from other place) as well, but it turned out, that the fragmentation code re-uses the first fragment packet as a base for the reassembled one, so the more flag (for Ipv6) or frag offset (for IPv6) would be only on those reassembled packets (and only on them).

krish2718 replied:
Ah, I thought those flags would be reset as once re-assembled they are of no use, but https://datatracker.ietf.org/doc/html/rfc791 doesn't say so, so, I guess it should work.

But this also means if someone uses that helper on each-fragment then it would falsely returns true, encording `fragment_offset == 0 && more_frags == 1` might mitigate this, then its a question of first fragment or reassembled packet.

rlubos replied:
I can add an explicit flag to `net_pkt` structure to indicate that packet is IP reassembled. There should be no ambiguity then.

krish2718 replied:
Ah, so, the L2 driver should set this flag? If yes, that should work, as the stack has no way of knowing its 1st fragment or re-assembled frame.

rlubos replied:
No, L2 has nothing to do with fragmentation. The flag would be off by default, and only set if the IP layer reassembled the packet.

krish2718 replied:
Ok, I guess we needn't worry about frag-reassembly offload.

krish2718 commented on tests/net/checksum_offload/src/main.c:
```diff
@@ -204,6 +271,61 @@ static void test_receiving(struct net_pkt *pkt)
+		}
+
+		/* Fragmented packet should have checksum set regardless of
+		 * chksum offloading
```
checksum

rlubos commented:
@krish2718 I've pushed an updated version.

rlubos committed:
net: pkt: Add explicit flag to indicate packet is IP reassembled

The current logic to determine whether a packet is IP reassembled is
flawed, as it only worked in certain conditions (which was ok, as the
conditions were satisfied for the current use case, but now it's a
public function). Therefore, add an explicit flag that indicates whether
a packet is IP reassembled or not.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
@jukkar @krish2718 Added one extra commit on top, adding an explicit flag to indicate that the packet has been IP reassembled (instead of using some ambiguous logic).

pdgendt commented on include/zephyr/net/net_pkt.h:
```diff
@@ -194,7 +194,12 @@ struct net_pkt {
+				  * the packet.
+				  */
+#if defined(CONFIG_NET_IPV4_FRAGMENT) || defined(CONFIG_NET_IPV6_FRAGMENT)
+	uint8_t ip_reassembled : 1; /* Packet is an reassembled IP packet. */
```
Nit
```suggestion
	uint8_t ip_reassembled : 1; /* Packet is a reassembled IP packet. */
```

pdgendt commented on include/zephyr/net/net_pkt.h:
```diff
@@ -822,6 +838,35 @@ static inline void net_pkt_set_ipv6_fragment_id(struct net_pkt *pkt,
 }
 #endif /* CONFIG_NET_IPV6_FRAGMENT */
 
+#if defined(CONFIG_NET_IPV4_FRAGMENT) || defined(CONFIG_NET_IPV6_FRAGMENT)
```
**A more general remark, definitely not something this PR should do;**

The code base has a lot of `#if defined(CONFIG_NET_IPV4_<feature>) || defined(CONFIG_NET_IPV6_<feature>)`,
I think we could use a hidden Kconfig to our advantage here:

```cfg
config NET_IP_FRAGMENT
	bool
	default y
	depends on NET_IPV4_FRAGMENT || NET_IPV6_FRAGMENT
```

And use this option in code:

```c
#ifdef CONFIG_NET_IP_FRAGMENT
/* Insert code */
#endif
```

Disclaimer: This is the lazy me, who wants to type/copy less, talking.

rlubos replied:
Good point actualy, I've added one more commit introducing the hidden symbol for IP fragmentation.

rlubos committed:
net: ip: Add hidden Kconfig symbol for IP fragmentation

Instead of consistently checking for both, IPv4 and IPv6 fragmentation
in several places, add a hidden Kconfig symbol which indicates that some
IP fragmentation has been enabled (either IPv4 or IPv6 or both).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>
